cloud/scope/powervs_cluster_test.go

```diff
@@ -3830,3 +3832,799 @@ func TestCreateVPCSubnet(t *testing.T) {
+			},
+			ResourceClient: mockResourceController,
+		}
+		mockResourceController.EXPECT().GetResourceInstance(gomock.Any()).Return(nil, nil, fmt.Errorf("error getting resource instance"))
+		err := clusterScope.DeleteCOSInstance()
+		g.Expect(err.Error()).To(Equal("failed to fetch COS service instance: error getting resource instance"))
+	})
+	t.Run("When cos instance state is active and DeleteResourceInstance succeeds", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+
+		clusterScope := PowerVSClusterScope{
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					COSInstance: &infrav1beta2.ResourceReference{
+						ID:                ptr.To("cosInstanceID"),
+						ControllerCreated: ptr.To(true),
+					},
+				},
+			},
+			ResourceClient: mockResourceController,
+		}
+		cosInstance := &resourcecontrollerv2.ResourceInstance{ID: ptr.To("cosInstanceID"), State: ptr.To(string(infrav1beta2.ServiceInstanceStateActive))}
+		mockResourceController.EXPECT().GetResourceInstance(gomock.Any()).Return(cosInstance, nil, nil)
+		mockResourceController.EXPECT().DeleteResourceInstance(gomock.Any()).Return(nil, nil)
+		err := clusterScope.DeleteCOSInstance()
+		g.Expect(err).To(BeNil())
+	})
+
+	t.Run("When cos instance state is active and DeleteResourceInstance returns error", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+
+		clusterScope := PowerVSClusterScope{
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+				Status: infrav1beta2.IBMPowerVSClusterStatus{
+					COSInstance: &infrav1beta2.ResourceReference{
+						ID:                ptr.To("cosInstanceID"),
+						ControllerCreated: ptr.To(true),
+					},
+				},
+			},
+			ResourceClient: mockResourceController,
+		}
+		cosInstance := &resourcecontrollerv2.ResourceInstance{ID: ptr.To("cosInstanceID"), State: ptr.To(string(infrav1beta2.ServiceInstanceStateActive))}
+		mockResourceController.EXPECT().GetResourceInstance(gomock.Any()).Return(cosInstance, nil, nil)
+		mockResourceController.EXPECT().DeleteResourceInstance(gomock.Any()).Return(nil, fmt.Errorf("error deleting resource instance"))
+		err := clusterScope.DeleteCOSInstance()
+		g.Expect(err).To(Equal(fmt.Errorf("error deleting resource instance")))
+	})
+}
+
+func TestDeleteServiceInstance(t *testing.T) {
+	var (
+		mockResourceController *mockRC.MockResourceController
+		mockCtrl               *gomock.Controller
+	)
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockResourceController = mockRC.NewMockResourceController(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When PowerVS service instance resource is not created by controller", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+
+		clusterScope := PowerVSClusterScope{IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{}}
+		isDeleted, err := clusterScope.DeleteServiceInstance()
```

Comment:
```suggestion
		requeue, err := clusterScope.DeleteServiceInstance()
```
Please change wherever applicable

Reply:
Done